contracts/BoostedMasterChefJoe.sol

```diff
@@ -0,0 +1,621 @@
+        pool.totalVeJoe = pool.totalVeJoe.sub(user.veJoeBalance).add(VEJOE.balanceOf(msg.sender));
+        user.veJoeBalance = VEJOE.balanceOf(msg.sender);
+
+        // Update the `pool.totalBoostedAmount`.
+        uint256 userBoost = getUserBoost(pid, msg.sender);
+        if (user.amount == 0) {
+            pool.totalBoostedAmount = pool.totalBoostedAmount.add(amount.mul(userBoost).div(1000));
+        } else {
+            pool.totalBoostedAmount = pool.totalBoostedAmount.sub(user.amount.mul(userBoost).div(1000)).add(
+                user.amount.add(amount).mul(userBoost).div(1000)
+            );
+        }
+
+        user.amount = user.amount.add(receivedAmount);
+        user.rewardDebt = user
+            .amount
+            .sub(amount)
+            .mul(getUserBoost(pid, msg.sender))
+            .mul(pool.accJoePerShare)
+            .div(1000)
+            .div(ACC_TOKEN_PRECISION);
+
+        // Interactions
+        IRewarder _rewarder = pool.rewarder;
+        if (address(_rewarder) != address(0)) {
+            _rewarder.onJoeReward(msg.sender, user.amount);
+        }
+        emit Deposit(msg.sender, pid, receivedAmount);
+    }
+
+    /// @notice Withdraw LP tokens from BMCJ.
+    /// @param pid The index of the pool. See `poolInfo`.
+    /// @param amount LP token amount to withdraw.
+    function withdraw(uint256 pid, uint256 amount) external nonReentrant {
+        harvestFromMasterChef();
+        updatePool(pid);
+        PoolInfo storage pool = poolInfo[pid];
+        UserInfo storage user = userInfo[pid][msg.sender];
+        uint256 userBoost = getUserBoost(pid, msg.sender);
+
+        if (user.amount > 0) {
+            // Harvest JOE
+            // This is split into two figures to avoid stack too deep error.
+            uint256 boostedAmount = user.amount.mul(getUserBoost(pid, msg.sender)).div(1000);
+
+            uint256 pending = boostedAmount.mul(pool.accJoePerShare).div(ACC_TOKEN_PRECISION).sub(user.rewardDebt).add(
+                claimableJoe[pid][msg.sender]
+            );
+
+            claimableJoe[pid][msg.sender] = 0;
+            JOE.safeTransfer(msg.sender, pending);
+            emit Harvest(msg.sender, pid, pending);
+        }
+
+        // Effects
+
+        // If the user has left the pool altogether we subtract all their
+        // veJoe balance from `pool.totalVeJoe`, and full boosted balance
+        // from `pool.totalBoostedAmount` otherwise we just adjust down the
+        // total boosted amount as total vejoe is unchanged.
+
+        // Update the `pool.totalBoostedAmount` and `totalVeJoe` if needed.
+        if (user.amount.sub(amount) == 0) {
+            pool.totalVeJoe = pool.totalVeJoe.sub(VEJOE.balanceOf(msg.sender));
+            pool.totalBoostedAmount = pool.totalBoostedAmount.sub(user.amount.mul(userBoost).div(1000));
+        } else {
+            pool.totalBoostedAmount = pool.totalBoostedAmount.sub(amount.mul(userBoost).div(1000));
```

Comment:
i think we can extract these two statements to outside of the if block, i.e. just do:
```
if (user.amount.sub(amount) == 0) { ... }
pool.totalBoostedAmount = pool.totalBoostedAmount.sub(amount.mul(userBoost).div(1000));
```